### PR TITLE
osutil/disks: workaround to map mapper devices with their underlying disk

### PR DESCRIPTION
**Pull request by sespiros**
Trying to match a mountpoint with its underlying disk currently fails if the mountpoint is for a mapper virtual device. This happens because diskFromPartUDevProps is using the ID_PART_ENTRY_DISK udev property to retrieve the MINOR and MAJOR IDs of the partition and that field is not present for mapper devices.

The extra use case that this is needed for is for partitions mounted with verity by mount. When attempting to mount a disk under /dev with verity options, mount will create automatically a mapper device under /dev/mapper following the naming convention <partition_name> -> libmnt_<partition name>. Rather than covering all the different cases (LVM, CRYPT partitions etc.), this workaround will simply translate a mapper device name to its physical device using the above naming convention.

These APIs are not meant to be called for tmpfs based mountpoints but they are excluded anyway from the workaround. This for example covers virtual devices that get auto-created by mount under i.e /dev/mapper/libmnt_tmpfs.

That is for https://warthogs.atlassian.net/browse/FR-8353

### CODE REVIEW

File: osutil/disks/disks_linux.go [hunk 1]
```diff
@@ -612,6 +613,17 @@ func partitionPropsFromMountPoint(mountpoint string) (source string, props map[s
 		return "", nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
 	}
 
+	// for verity partitions, mount creates /dev/mapper/libmnt_<partition> devices
```
Comment (valentindavid) [inline osutil/disks/disks_linux.go:616]: That works only if we use `verity.` mount options. Not if we move to use `veritysetup open` for example.

Could you find it in the device slaves in sysfs?

File: osutil/disks/disks_linux.go [hunk 2]
```diff
@@ -612,6 +613,17 @@ func partitionPropsFromMountPoint(mountpoint string) (source string, props map[s
...
+	// automatically. udevPropertiesForName fails to detect partition information
+	// for mapper devices therefore if one of these is encountered, it is translated
+	// to the underlying physical device.
+	r := `^/dev/mapper/libmnt_(\w+)$`
+	re := regexp.MustCompile(r)
```
[the remark this answers is not in the document]
Comment (sespiros) [reply osutil/disks/disks_linux.go:621]: I re-implemented this with HasPrefix, StripPrefix for now as suggested. Thanks for the recommendation :) I will revisit though if I think of a use-case that cannot properly be handled like that.

File: osutil/disks/disks_linux.go (hunk 1 above)
Comment (sespiros) [reply osutil/disks/disks_linux.go:615]: Correct this should only work with how we are mounting verity partitions today, through `doSystemdMount`.

I am suggesting this as a workaround as I figured that to properly solve this, we would need to come up with more complicated logic that maps the various potential types of mapper devices back to physical disks. Is there interest for this?

Also I haven't exhaustively researched this and there might be a quicker way to reach to the same result somehow through sysfs but I based this on top of how we already do this using `udevadm`.

This is the process I followed as an example:
To map a mapper device for a verity partition that was mounted like:
```bash
veritysetup open /dev/vda1 rootfs /dev/vda2 <root-hash>
```
back to its corresponding physical disk, a path would be:

1. first do a call to  `veritysetup status` to retrieve the data device:
```bash
$ veritysetup status /dev/mapper/rootfs
  type:        VERITY
  ...
  data device: /dev/nbd0p1
  ...
```

2. Then invoke `udevadm info` as we already do:
```bash
udevadm info --query property --name /dev/nbd0p1
```

But there could be other types such as "LVM". For my system's disk:
```bash
udevadm info --query property --name /dev/mapper/vgubuntu-root
```
returns output that is also lacking the `ID_PART_ENTRY__UID` which we are using to compare partitions. 

For this partition the process would differ from the verity one as we would need to:
1. First identify that this is an LVM partition and retrieve the volume group:
```bash
$ sudo lvdisplay /dev/mapper/vgubuntu-root
  --- Logical volume ---
...
  VG Name                vgubuntu
...
```
2. Identify the underlying physical volume maybe with something like:
```bash
$ sudo pvs | grep vgubuntu
  /dev/mapper/nvme0n1p3_crypt vgubuntu lvm2 a--  <930.28g    0
```
3. Then, in that case it is a crypt device so we can get to the physical device with:
```bash
$ sudo cryptsetup status /dev/mapper/nvme0n1p3_crypt
/dev/mapper/nvme0n1p3_crypt is active and is in use.
...
  device:  /dev/nvme0n1p3
...
```
4. and finally call `udevadm` to retrieve the `ID_PART_ENTRY_UUID`:
```bash
$udevadm info --query property --name /dev/nvme0n1p3
```

**Review by zyga — COMMENTED** [review]
As much as generic coding is concerned this looks fine. I will refrain from accepting as I would need a deep dive into the storage layer to find potential issues.

File: osutil/disks/disks_linux.go (hunk 1 above)
Comment (valentindavid) [reply osutil/disks/disks_linux.go:615]: If we have to call lots of commands like that, I would prefer to call them in udev rules and populate variables for the devices. So we can just read it from the properties. And maybe those rules would be good candidate to submit upstream to systemd.

Comment (valentindavid) [reply osutil/disks/disks_linux.go:615]: Like a small program that links to libcryptsetup and call `crypt_get_verity_info` for the `DEVNAME`. Dumps the output as udev format. Then call it in a rule with `IMPORT{program}`.

Do similar thing for lvm (I suppose there is a library for it, but I do not know it).

Then if that works, try to make udev builtins of them in systemd.

Comment (sespiros) [reply osutil/disks/disks_linux.go:615]: I opened https://github.com/snapcore/snapd/pull/14236 to try and solve this in a more universal way.

**Comment by sespiros** [comment]
@zyga @valentindavid we can reject this one if the solution in https://github.com/snapcore/snapd/pull/14236 is preferred.

**Pull request closed without merging**